### ISSUE TITLE
Query to access usage of usage of Critical CSS

### ISSUE THREAD

**Issue by kt-12** [GoogleChromeLabs/wpp-research#16]
* Create HTTP Archive query to get % of WordPress sites that do not use Critical CSS.

**Comment by kt-12** [comment]
To get sites with style in head
```sql
SELECT * FROM `httparchive.response_bodies.2022_10_01_mobile` 
where
REGEXP_CONTAINS(
    body, 
    '<head.+<style(.+)/style>.+/head>'
)
AND page=url
limit 50
```

**Comment by felixarntz** [comment]
@kt-12 With this query consuming ~154 TB of data, running this will be extremely expensive and time-consuming. (A query I ran the other day that consumed ~7 TB already took almost 40 minutes.)

It's amazing you were able to come up with that query though! Is there maybe a way to make it use only the monthly tables (using `2022_10_01` is fine). That should result in a notable improvement. Can you use the monthly `pages`, `requests`, and `technologies` tables instead?

**Comment by kt-12** [comment]
@felixarntz  I can change the query to use monthly tables. I'll update it here once done.

**Comment by kt-12** [comment]
New query scan's 73TB. The primary issue is that I have to scan the head of each page's HTML for inline style in the head from `response_bodies` table.  We have metrics for inline style on the entire page, but I couldn't find anything specific to the head.


```sql
CREATE TEMP FUNCTION
  getRenderBlockingCSSCount(lighthouse STRING)
  RETURNS INT64
  LANGUAGE js AS '''
try {
const report = JSON.parse(lighthouse);
const networkItems = report.audits["network-requests"].details.items
const renderBlockingCss = networkItems.filter(item => item.resourceType == "Stylesheet" && item.priority == "VeryHigh");
return renderBlockingCss.length
} catch (e) {
return -1;
}
''';
SELECT
  client,
  total_wp_sites,
  sites_with_critical_css,
  (total_wp_sites-sites_with_critical_css) AS sites_without_critical_css,
  CONCAT(ROUND((total_wp_sites-sites_with_critical_css)*100/total_wp_sites, 3),' %') AS opportunity
FROM (
  SELECT
    lh._TABLE_SUFFIX AS client,
    COUNT(lh.url) AS total_wp_sites,
    COUNTIF( getRenderBlockingCSSCount(lh.report) = 0
      AND IFNULL( REGEXP_CONTAINS( res.body, '<head.+<style(.+)/style>.+/head>' ), FALSE ) ) AS sites_with_critical_css
  FROM
    `httparchive.lighthouse.2022_10_01_*` AS lh
  JOIN
    `httparchive.technologies.2022_10_01_*` AS tech
  ON
    tech.url = lh.url
  JOIN
    `httparchive.response_bodies.2022_10_01_*` AS res
  ON
    res.page = lh.url
  WHERE
    app = 'WordPress'
    AND category = 'CMS'
    AND lh._TABLE_SUFFIX = tech._TABLE_SUFFIX
    AND lh._TABLE_SUFFIX = res._TABLE_SUFFIX
    AND REGEXP_CONTAINS(JSON_EXTRACT(lh.report,'$.audits.network-requests.details.items' ), '"priority"')
  GROUP BY
    lh._TABLE_SUFFIX )
```

**Comment by felixarntz** [comment]
@kt-12 That last query looks great to me, as far as I understand it. A few follow up questions for clarification:
* Is `$._renderBlockingCSS` only for render blocking _external_ CSS?
* Does `$._inline_style_bytes` cover inline `<style>` tags in the entire document or only in `<head>`?

Maybe the second question isn't really that important; if the site has zero render-blocking external CSS, there's a good chance it is inlining critical CSS. At least some of the inline `<style>` tags are then probably in the `<head>`, otherwise there would be no style rules in `<head>` at all, which I cannot imagine.

However, one potential concern is that the query could also include sites that simply inline _all_ of their CSS, not just the critical CSS (for example that is a requirement in AMP documents). I'm not sure there's a way to avoid this, but maybe we should add another condition to only count sites that _do_ have another external stylesheet somewhere _after_ `<head>` (anywhere in the `<body>`)? That makes it more likely that they use critical CSS inlining.

Another thing, can you please put the query in a pull request? It's not great from a workflow perspective to discuss them in an issue, for example I can't comment on specific lines. It's perfectly fine to have a draft PR, and if then you come up with another query, open another draft PR. Eventually we will get there.

Overall this last query looks quite promising.

**Comment by kt-12** [comment]
@felixarntz  Thank you for the review. 

I am addressing your questions below:

* Yes, `$._renderBlockingCSS` only keeps a count of render-blocking external CSS. 
* `$._inline_style_bytes`, as per my knowledge, keeps track of style on the entire page and not just the `<head>`; I'll still need to confirm this.  

I agree with you on your concern about the AMP pages. There is an attribute available in the same payload under `_markup`, `amp.amp_page`, which is set to true if it is an AMP page. I'll add that to my query to filter such cases.

Traversing an entire page for `<style>`  in `<head>` and `<body>` has so far been expensive. Given the opportunity of 95% obtained from the current query, there is little we can filter out from this activity.

I have also moved my query to the draft PR. Since this is the only query that could run in a few minutes, I replaced the old query in PR and only kept this one.
